lib/langfuse/stale_while_revalidate.rb

```
@@ -0,0 +1,262 @@
+    # @return [Object] Cached, stale, or freshly fetched value
+    # @raise [ConfigurationError] if SWR is not enabled (stale_ttl is not positive)
+    #
+    # @example
+    #   cache.fetch_with_stale_while_revalidate("greeting:v1") do
+    #     api_client.get_prompt("greeting")
+    #   end
+    def fetch_with_stale_while_revalidate(key, &)
+      raise ConfigurationError, "fetch_with_stale_while_revalidate requires a positive stale_ttl" unless swr_enabled?
+
+      entry = cache_get(key)
+
+      if entry&.fresh?
+        # FRESH - return immediately
+        logger.debug("CACHE HIT!")
+        entry.data
+      elsif entry&.stale?
+        # REVALIDATE - return stale + refresh in background
+        logger.debug("CACHE STALE!")
+        schedule_refresh(key, &)
+        entry.data # Instant response!
+      else
+        # MISS - must fetch synchronously
+        logger.debug("CACHE MISS!")
+        fetch_and_cache(key, &)
+      end
+    end
+
+    # Check if SWR is enabled
+    #
+    # SWR is enabled when stale_ttl is positive, meaning there's a grace period
+    # where stale data can be served while revalidating in the background.
+    #
+    # @return [Boolean] true if stale_ttl is positive
+    def swr_enabled?
+      stale_ttl.positive?
+    end
+
+    # Shutdown the cache refresh thread pool gracefully
+    #
+    # @return [void]
+    def shutdown
+      return unless @thread_pool
+
+      @thread_pool.shutdown
+      @thread_pool.wait_for_termination(5) # Wait up to 5 seconds
+    end
+
+    private
+
+    # Initialize thread pool for background refresh operations
+    #
+    # @param refresh_threads [Integer] Maximum number of refresh threads
+    # @return [Concurrent::CachedThreadPool]
+    def initialize_thread_pool(refresh_threads)
+      Concurrent::CachedThreadPool.new(
+        max_threads: refresh_threads,
+        min_threads: 2,
+        max_queue: 50,
+        fallback_policy: :discard # Drop oldest if queue full
```

On `fallback_policy: :discard`

Concern: The comment says "Drop oldest if queue full" but `:discard` actually drops the newest task
(the one being submitted), not the oldest. The comment is incorrect.

Additionally: I agree `:discard` is  the right policy here: when the queue is full, we silently drop the
refresh request. The user gets stale data (fine), but they have no visibility that refreshes are
being dropped.

Two suggestions:

```
fallback_policy: :discard # Silently drop refresh if queue full (stale data continues to be served)
```

Or if we want visibility:
```
fallback_policy: ->(executor, task) {
  logger.warn("SWR refresh queue full, dropping refresh task")
}
```

I'm in favor of just updating the comment, but wanted to share my thinking here. 


```suggestion
        fallback_policy: :discard # Silently drop refresh if queue full (stale data continues to be served)
```